pkg/metallb/frr.go

```diff
@@ -0,0 +1,559 @@
+
+	return builder
+}
+
+// WithBGPNeighbor defines a single neighbor IP and ASN number.
+func (builder *FrrConfigurationBuilder) WithBGPNeighbor(bgpPeerIP string, remoteAS uint32) *FrrConfigurationBuilder {
+	if valid, _ := builder.validate(); !valid {
+		return builder
+	}
+
+	glog.V(100).Infof(
+		"Define BGP Neighbor %s with peer IP %s and ASN %v",
+		builder.Definition.Name, bgpPeerIP, remoteAS)
+
+	if bgpPeerIP == "" {
+		glog.V(100).Infof("The BGP peer IP should not be blank")
+
+		builder.errorMsg = fmt.Sprintf("the BGP peer IP %s is an empty string", bgpPeerIP)
+	}
+
+	if net.ParseIP(bgpPeerIP) == nil {
+		glog.V(100).Infof("The peerIP for the frrConfiguration bgp peer contains invalid ip address %s",
+			bgpPeerIP)
+
+		builder.errorMsg = "frrConfiguration 'peerIP' of the BGPPeer contains invalid ip address"
+	}
+
+	if strconv.Itoa(int(remoteAS)) == "" {
+		glog.V(100).Infof("The remoteASN of the frrConfiguration can not be empty")
+
+		builder.errorMsg = "frrConfiguration 'remoteAS' cannot be empty"
+	}
+
+	if builder.errorMsg != "" {
+		return builder
+	}
+
+	if len(builder.Definition.Spec.BGP.Routers[0].Neighbors) == 0 {
+		builder.Definition.Spec.BGP.Routers[0].Neighbors = []frrtypes.Neighbor{{
+			Address: bgpPeerIP,
+			ASN:     remoteAS,
+		},
+		}
+	} else {
+		builder.Definition.Spec.BGP.Routers[0].Neighbors = append(builder.Definition.Spec.BGP.Routers[0].Neighbors,
+			frrtypes.Neighbor{
+				Address: bgpPeerIP,
+				ASN:     remoteAS,
+			})
+	}
+
+	return builder
+}
+
+// WithBGPPassword defines the password used between BGP peers to form adjacency and attaches the password using the
+// neighbor index from the neighbor list.
+func (builder *FrrConfigurationBuilder) WithBGPPassword(bgpPassword string,
+	neighborIndex int) *FrrConfigurationBuilder {
+	if valid, _ := builder.validate(); !valid {
+		return builder
+	}
+
+	glog.V(100).Infof(
+		"Define FRRConfiguration %s in namespace %s with the password: %s",
+		builder.Definition.Name, builder.Definition.Namespace, bgpPassword)
+
+	if bgpPassword == "" {
+		glog.V(100).Infof("The bgpPassword should not be blank")
+
+		builder.errorMsg = fmt.Sprintf("the bgpPassword %s is an empty string", bgpPassword)
+	}
+
+	if strconv.Itoa(neighborIndex) == "" {
+		glog.V(100).Infof("The neighborIndex cannot be blank when adding a BGP password")
+
+		builder.errorMsg = "the 'neighborIndex' cannot be empty when configuring a BGP password"
+	}
+
+	if builder.errorMsg != "" {
+		return builder
+	}
+
+	if len(builder.Definition.Spec.BGP.Routers[0].Neighbors) == 0 {
+		builder.Definition.Spec.BGP.Routers[0].Neighbors = []frrtypes.Neighbor{
+			{
+				Password: bgpPassword,
+			},
+		}
+	} else {
+		builder.Definition.Spec.BGP.Routers[0].Neighbors[neighborIndex].Password = bgpPassword
+	}
+
+	return builder
+}
+
+// WithEBGPMultiHop sets the EBGP multihop setting on the remote BGP peer.
+func (builder *FrrConfigurationBuilder) WithEBGPMultiHop(neighborIndex int) *FrrConfigurationBuilder {
```

Comment:
I think we should get RouterIndex for all related funcs as well .

Reply:
I kept the router hard coded as 0.  The currently is no plans to incorportate VRFs into frr.